test/reparent.py

```diff
@@ -621,6 +621,86 @@ def test_reparent_with_down_slave(self, shard_id='0'):
     tablet.kill_tablets([tablet_62344, tablet_62044, tablet_41983,
                          tablet_31981])
 
+  def test_change_type_semi_sync(self):
+    utils.run_vtctl(['CreateKeyspace', 'test_keyspace'])
+
+    # create the database so vttablets start, as they are serving
+    tablet_62344.create_db('vt_test_keyspace')
+    tablet_62044.create_db('vt_test_keyspace')
+    tablet_41983.create_db('vt_test_keyspace')
+    tablet_31981.create_db('vt_test_keyspace')
+
+    # Start up a soon-to-be master, one replica and two rdonly.
+    tablet_62344.init_tablet('replica', 'test_keyspace', '0', start=True,
+                             wait_for_start=False)
+    tablet_62044.init_tablet('replica', 'test_keyspace', '0', start=True,
+                             wait_for_start=False)
+    tablet_41983.init_tablet('rdonly', 'test_keyspace', '0', start=True,
+                             wait_for_start=False)
+    tablet_31981.init_tablet('rdonly', 'test_keyspace', '0', start=True,
+                             wait_for_start=False)
+    for t in [tablet_62344, tablet_62044, tablet_41983, tablet_31981]:
+      t.wait_for_vttablet_state('NOT_SERVING')
+
+    # Force the slaves to reparent assuming that all the datasets are
+    # identical.
+    utils.run_vtctl(['InitShardMaster', '-force', 'test_keyspace/0',
+                     tablet_62344.tablet_alias], auto_log=True)
+    utils.validate_topology(ping_tablets=True)
+    self._check_master_tablet(tablet_62344)
+
+    # Stop replication on rdonly1, to make sure when we make it
+    # replica it doesn't start again.
+    # Note we do a similar test for replica -> rdonly below.
+    utils.run_vtctl(['StopSlave', tablet_41983.tablet_alias])
+
+    # Check semi-sync on slaves.
+    # The flag is only an indication of the value to use next time
+    # we turn replication on, so also check the status.
+    # tablet_41983 is not replicating, so its status is off.
+    tablet_62044.check_db_var('rpl_semi_sync_slave_enabled', 'ON')
+    tablet_41983.check_db_var('rpl_semi_sync_slave_enabled', 'OFF')
+    tablet_31981.check_db_var('rpl_semi_sync_slave_enabled', 'OFF')
+    tablet_62044.check_db_status('rpl_semi_sync_slave_status', 'ON')
+    tablet_41983.check_db_status('rpl_semi_sync_slave_status', 'OFF')
+    tablet_31981.check_db_status('rpl_semi_sync_slave_status', 'OFF')
+
+    # Change replica to rdonly while replicating, should turn off semi-sync,
+    # and restart replication.
+    utils.run_vtctl(['ChangeSlaveType', tablet_62044.tablet_alias, 'rdonly'],
+                    auto_log=True)
+    tablet_62044.check_db_var('rpl_semi_sync_slave_enabled', 'OFF')
+    tablet_62044.check_db_status('rpl_semi_sync_slave_status', 'OFF')
+
+    # Change rdonly1 to replica, should turn on semi-sync, and not start rep.
```

Comment:
This test is very hard to follow now with all these numerically-named tablets. It would be great if they had better aliases, like tablet_replica, tablet_rdonly1, tablet_rdonly2, etc.